quinn-proto/src/connection/mod.rs

```diff
@@ -1749,20 +1742,18 @@ where
         if end > max {
             return Err(TransportError::CRYPTO_BUFFER_EXCEEDED(""));
         }
+
         space
             .crypto_stream
             .insert(crypto.offset, crypto.data.clone());
-        let mut buf = [0; 8192];
-        loop {
-            let n = space.crypto_stream.read(&mut buf).unwrap();
-            if n == 0 {
-                return Ok(());
-            }
-            trace!("consumed {} CRYPTO bytes", n);
-            if self.crypto.read_handshake(&buf[..n])? {
+        while let Some(buf) = space.crypto_stream.read(usize::MAX).unwrap() {
+            trace!("consumed {} CRYPTO bytes", buf.len());
+            if self.crypto.read_handshake(&buf)? {
```

Comment:
What if the crypto stream doesn't consume all bytes (e.g. because they might only be required in a later phase?). IT seems like those are lost then. Is it guaranteed that this won't happen.

Reply:
`read_handshake` must buffer internally; there's no way for it to express incomplete reads, and this is consistent with [the behavior of rustls](https://docs.rs/rustls/0.19.0/rustls/quic/trait.QuicExt.html#tymethod.read_hs).